pkg/metallb/frr.go

```diff
@@ -0,0 +1,589 @@
+package metallb
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	frrtypes "github.com/openshift-kni/eco-goinfra/pkg/schemes/metallb/frrtypes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtimeClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// FrrConfigurationBuilder provides struct for the FrrConfiguration object containing connection to
+// the cluster and the FrrConfiguration definitions.
+type FrrConfigurationBuilder struct {
+	Definition *frrtypes.FRRConfiguration
+	Object     *frrtypes.FRRConfiguration
+	apiClient  runtimeClient.Client
+	errorMsg   string
+}
+
+// NewFrrConfigurationBuilder creates a new instance of FRRConfiguration.
+func NewFrrConfigurationBuilder(
+	apiClient *clients.Settings, name, nsname string) *FrrConfigurationBuilder {
+	glog.V(100).Infof(
+		"Initializing new Frrconfiguration structure with the following params: %s, %s",
+		name, nsname)
+
+	if apiClient == nil {
+		glog.V(100).Infof("failed to initialize the apiclient is empty")
+
+		return nil
+	}
+
+	err := apiClient.AttachScheme(frrtypes.AddToScheme)
+	if err != nil {
+		glog.V(100).Infof("Failed to add metallb scheme to client schemes")
+
+		return nil
+	}
+
+	builder := FrrConfigurationBuilder{
+		apiClient: apiClient,
+		Definition: &frrtypes.FRRConfiguration{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the frrConfiguration is empty")
+
+		builder.errorMsg = "frrConfiguration 'name' cannot be empty"
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the frrConfiguration is empty")
+
+		builder.errorMsg = "frrConfiguration 'nsname' cannot be empty"
+	}
+
+	return &builder
+}
+
+// Create makes a FrrConfiguration in the cluster and stores the created object in struct.
+func (builder *FrrConfigurationBuilder) Create() (*FrrConfigurationBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating the frrconfiguration %s in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace,
+	)
+
+	var err error
+	if !builder.Exists() {
+		err = builder.apiClient.Create(context.TODO(), builder.Definition)
+
+		if err != nil {
+			glog.V(100).Infof("Failed to create MetalLb")
+
+			return nil, err
+		}
+
+		builder.Object = builder.Definition
+	}
+
+	return builder, err
```

Comment:
I'd rather be explicit about returning nil after a successful run similar to your Create and Delete.

```suggestion
	return builder, nil
```

Reply:
replaced